### PR TITLE
Add support for io.Writer/Reader/Seeker in VFSfh

### PR DESCRIPTION
**Pull request by Shelnutt2**
This allows a VFS file handle to be used in functions which take one of the io package interfaces

### CODE REVIEW

File: vfs.go [hunk 1]
```diff
@@ -396,3 +399,107 @@ func (v *VFS) DirSize(uri string) (uint64, error) {
...
+		}
+		v.size = &size
+	}
+
+	// If the requested read size is beyond the limit, adjust bytes to read
+	if v.offset+nbytes > *v.size {
+		nbytes = *v.size - v.offset
+	}
+
+	if nbytes == 0 {
```
Comment (antalakas) [inline vfs.go:436]: Why not check this first and exit early?

File: vfs.go [hunk 2]
```diff
@@ -396,3 +399,107 @@ func (v *VFS) DirSize(uri string) (uint64, error) {
...
+
+	v.offset += nbytes
+	return int(nbytes), nil
+}
+
+// Write the contents of a buffer into a file. Note that this function only
+// appends data at the end of the file. If the file does not exist,
+// it will be created
+func (v *VFSfh) Write(bytes []byte) (int, error) {
+	cbuffer := C.CBytes(bytes)
```
Comment (antalakas) [inline vfs.go:455]: `C.CBytes ` or `unsafe.Pointer(&bytes[0])`

File: vfs.go [hunk 3]
```diff
@@ -396,3 +399,107 @@ func (v *VFS) DirSize(uri string) (uint64, error) {
...
+
+	switch whence {
+	case io.SeekStart:
+		origin = 0
+		break
+	case io.SeekCurrent:
+		origin = v.offset
+		break
+	case io.SeekEnd:
+		origin = *v.size
```
Comment (antalakas) [inline vfs.go:492]: Are we sure that `v.size` is set?

Comment (antalakas) [reply vfs.go:492]: Also origin is 0 after `var origin uint64` so `SeekEnd` will be 0 always right?

**Review by antalakas — CHANGES_REQUESTED** [review]

File: vfs.go (hunk 1 above)
Comment (Shelnutt2) [reply vfs.go:436]: We can't check this first incase we need to set the nbytes from the size - offset.

File: vfs.go (hunk 3 above)
Comment (Shelnutt2) [reply vfs.go:492]: Thanks, adjusted

**Review by snagles — APPROVED** [review]

File: vfs.go [hunk 4]
```diff
@@ -462,3 +465,112 @@ func (v *VFS) NumOfFragmentsInPath(path string) (int, error) {
...
+	return nil
+}
+
+// Read part of a file
+func (v *VFSfh) Read(p []byte) (int, error) {
+	nbytes := uint64(len(p))
+
+	// If the size is empty, fetch it
+	if v.size == nil {
+		size, err := v.vfs.FileSize(v.uri)
```
Comment (antalakas) [inline vfs.go:490]: This happens in two places, maybe a wrapper function is a good choice

File: vfs.go [hunk 5]
```diff
@@ -462,3 +465,112 @@ func (v *VFS) NumOfFragmentsInPath(path string) (int, error) {
...
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Unknown error in VFS.Close: %s", v.context.LastError())
+	}
+
+	v.Free()
+	return nil
+}
+
+// Read part of a file
+func (v *VFSfh) Read(p []byte) (int, error) {
```
Comment (antalakas) [inline vfs.go:485]: Read and Write signatures end up the same and this confuses me. I don't know if the out type argument is used in go, but i 'd prefer to allocate memory in this function taking the number of bytes needed and returning the []byte

**Review by antalakas — APPROVED** [review]

File: vfs.go (hunk 5 above)
Comment (Shelnutt2) [reply vfs.go:485]: We can't edit the Read/Write signatures they are part of the interface(s): https://golang.org/pkg/io/#Reader

File: vfs.go (hunk 4 above)
Comment (Shelnutt2) [reply vfs.go]: Good idea

**Pull request merged**